전체 영화목록, 상영중인 영화목록 불러오기#26

nahokyun opened the pull request:
> ### 추가 및 변경된점
- application.properties에 moviePerPage값을 지정하여 MovieService.java의 생성자부분을 변경하였습니다.
- 영화와 포스터 정보를 담을 MovieAndPosterVO를 추가하였습니다.
- movieService에 입력받은 현재페이지넘버와 status를 통해 전체영화목록인지 현재상영중인 영화목록인지를 골라 리턴하는 getMovieAndPosterList메소드를 추가하였습니다.
- movieService의 메소드들에 javadoc을 추가하였습니다.

## MainController.java
```java
    @GetMapping("/get-movies-list")
    @ResponseStatus(HttpStatus.OK)
    public PagingVO getMovieAndPosterList(@RequestParam Integer currentPageNum,Boolean status){
        return movieService.getMovieAndPosterList(currentPageNum,status);
    }

```
## MovieAndPosterVO.java
```java
@Getter
@RequiredArgsConstructor
public class MovieAndPosterVO {
    private final Long movieId;
    private final String movieName;
    private final String genre;
    private final String actor;
    private final Integer runningTime;
    private final String supervisor;
    private final Date openingDate;
    private final String saveFilePath;
}
```
## MovieService.java
### 추가된 메소드
```java
    /**
     * 전체 영화목록 또는 현재 상영중인 영화(개봉일 1달 이내) 목록들을 불러오는 메소드
     *
     * @param currentPageNum 현재페이지를 나타내는 Integer 값
     * @param status 모든영화목록(1)을 가져올지 현재상영영화만 가져올지(0)에대한 Boolean 값
     * @return PagingVO 현재페이지에 해당하는 데이터들의 값을 PagingVO 데이터타입으로 반환
     */
    public PagingVO getMovieAndPosterList(Integer currentPageNum,Boolean status){

        int total;
        int totalPageNum;
        int startIdx = (currentPageNum - 1) * moviePerPage;
        List<MovieAndPosterVO> data;

        try{
            if(status.booleanValue()) {
                total = movieMapper.getTotalRowCount();
                data = movieMapper.getAllMovieAndPosterList(startIdx,moviePerPage);
            }
            else {
                total = movieMapper.getCurrentRowCount();
                data = movieMapper.getCurrentMovieAndPosterList(startIdx,moviePerPage);
            }

            totalPageNum = (int) Math.ceil((double) total/moviePerPage);
        } catch (DataAccessException e){
            throw  new IllegalArgumentException(FAILED_BRING_REVIEW.getMessage());
        }

        PagingVO result = PagingVO.builder()
                .total(total)
                .totalPageNum(totalPageNum)
                .currentPage(currentPageNum)
                .data(data)
                .build();

        return result;
    }

```

## MovieMapper.xml
### 추가된 쿼리문
- opening_date를 기준으로 1달 이내인 영화들은 현재 상영중인 영화목록에 들어가게됩니다.
```xml

    <select id="getCurrentRowCount" resultType="int">
        select count (*)
        from movie
        where opening_date between DATE_ADD (NOW(), interval -1 week) and NOW()
    </select>

    <select id="getAllMovieAndPosterList" resultType="MovieAndPosterVO">
        select movie.id, movie.movie_name, movie.genre, movie.actor, movie.running_time, movie.supervisor, movie.opening_date, poster.save_file_path
        from movie left join poster
        on movie.id = poster.movie_id
        order by movie.id Desc
        limit #{startIdx}, #{moviePerPage}
    </select>

    <select id="getCurrentMovieAndPosterList" resultType="MovieAndPosterVO">
        select movie.id, movie.movie_name, movie.genre, movie.actor, movie.running_time, movie.supervisor, movie.opening_date, poster.save_file_path
        from movie left join poster
        on movie.id = poster.movie_id
        where movie.opening_date between DATE_ADD (NOW(), interval -1 week) and NOW()
        order by movie.id Desc
        limit #{startIdx}, #{moviePerPage}
    </select>

```

nahokyun committed:
전체 영화목록, 상영중인 영화목록 불러오기#26

application.properties에 moviePerPage값을 지정하여 MovieService.java의 생성자부분을 변경하였습니다.

영화와 포스터 정보를 담을 MovieAndPosterVO를 추가하였습니다.

입력받은 현재페이지넘버와 status를 통해 전체영화목록인지 현재상영중인 영화목록인지를 골라 리턴하는 getMovieAndPosterList메소드를 추가하였습니다.

sunwootest commented on src/main/java/study/movieservice/controller/MainController.java:
```diff
@@ -28,4 +30,13 @@ public String posterJoin(@RequestPart MultipartFile file,@RequestPart Long movie
         return SUCCESS_SAVE_POSTER.getMessage();
     }
+
+    @GetMapping("/get-movies-list")
+    @ResponseStatus(HttpStatus.OK)
+    public PagingVO getMovieAndPosterList(@RequestParam Integer currentPageNum,Boolean status){
+
+        return movieService.getMovieAndPosterList(currentPageNum,status);
+    }
```
GetMappin 자체에 이미 get이라는 의미가 포함되어있습니다. url에 굳이 get을 추가로 붙일 필요는 없어보입니다.

sunwootest commented on src/main/java/study/movieservice/service/MovieService.java:
```diff
@@ -38,4 +65,43 @@ public void addPoster(MultipartFile file, Long movieId) {
+     * @param status 모든영화목록(1)을 가져올지 현재상영영화만 가져올지(0)에대한 Boolean 값
+     * @return PagingVO 현재페이지에 해당하는 데이터들의 값을 PagingVO 데이터타입으로 반환
+     */
+    public PagingVO getMovieAndPosterList(Integer currentPageNum,Boolean status){
+
+        int total;
+        int totalPageNum;
+        int startIdx = (currentPageNum - 1) * moviePerPage;
+        List<MovieAndPosterVO> data;
+
+        try{
+            if(status.booleanValue()) {
+                total = movieMapper.getTotalRowCount();
+                data = movieMapper.getAllMovieAndPosterList(startIdx,moviePerPage);
+            }
+            else {
+                total = movieMapper.getCurrentRowCount();
+                data = movieMapper.getCurrentMovieAndPosterList(startIdx,moviePerPage);
+            }
+
+            totalPageNum = (int) Math.ceil((double) total/moviePerPage);
+        } catch (DataAccessException e){
+            throw  new IllegalArgumentException(FAILED_BRING_REVIEW.getMessage());
+        }
+
+        PagingVO result = PagingVO.builder()
+                .total(total)
+                .totalPageNum(totalPageNum)
+                .currentPage(currentPageNum)
+                .data(data)
+                .build();
+
+        return result;
+    }
```
status를 int로 표현하는 것은 유지보수 측면에서 안좋아보이네요. enum을 활용해도 좋을 것 같습니다.

nahokyun committed:
전체 영화목록, 상영중인 영화목록 불러오기 #26

MainController에서 영화목록을 불러오는 url변경

status의 타입을 string으로 받아 enum형식으로 처리

nahokyun commented:
리뷰 반영했습니다!

> ### 변경된점
- MainController에서 영화목록을 불러오는 url을 변경하였습니다.
- status의 데이터타입을 boolean에서 string으로 입력받아 movieService에서 enum형식으로 변경하여 처리하였습니다.
- 영화목록의 종류를 나타내느 새로운 enum타입의 MovieListType이 추가되었습니다.
## MainController.java
### 변경 전 
```java
    @GetMapping("/get-movies-list")
    @ResponseStatus(HttpStatus.OK)
    public PagingVO getMovieAndPosterList(@RequestParam Integer currentPageNum,Boolean status){

        return movieService.getMovieAndPosterList(currentPageNum,status);
    }

```
### 변경 후 
```java
    @GetMapping("/movies-list")
    @ResponseStatus(HttpStatus.OK)
    public PagingVO getMovieAndPosterList(@RequestParam Integer currentPageNum, String status){

        return movieService.getMovieAndPosterList(currentPageNum,status);
    }

```
## MovieService.java
- controller layer에서 enum타입으로 받는것도 가능하였지만 비정상적인 enum값이 들어왔을때 발생하는 illegalArgumentException을 핸들링하기 위하여 service layer내부 try문에서 형변환을 위해 valueOf를 사용할때 해당 exception을 catch하여 enum값이 잘못되었다는 메시지를 throw하도록 구현하였습니다.
### 변경 전 
```java
    public PagingVO getMovieAndPosterList(Integer currentPageNum,Boolean status){

        int total;
        int totalPageNum;
        int startIdx = (currentPageNum - 1) * moviePerPage;
        List<MovieAndPosterVO> data;

        try{
            if(status.booleanValue()) {
                total = movieMapper.getTotalRowCount();
                data = movieMapper.getAllMovieAndPosterList(startIdx,moviePerPage);
            }
            else {
                total = movieMapper.getCurrentRowCount();
                data = movieMapper.getCurrentMovieAndPosterList(startIdx,moviePerPage);
            }

            totalPageNum = (int) Math.ceil((double) total/moviePerPage);
        } catch (DataAccessException e){
            throw  new IllegalArgumentException(FAILED_BRING_REVIEW.getMessage());
        }

        PagingVO result = PagingVO.builder()
                .total(total)
                .totalPageNum(totalPageNum)
                .currentPage(currentPageNum)
                .data(data)
                .build();

        return result;
    }

```
### 변경 후 
```java
    public PagingVO getMovieAndPosterList(Integer currentPageNum,String status){

        int total=0;
        int totalPageNum;
        int startIdx = (currentPageNum - 1) * moviePerPage;
        List<MovieAndPosterVO> data=null;

        try{
            MovieListType listStatus=MovieListType.valueOf(status);

            if(listStatus==ALL) {
                total = movieMapper.getTotalRowCount();
                data = movieMapper.getAllMovieAndPosterList(startIdx,moviePerPage);
            }
            else if(listStatus==CURRENT){
                total = movieMapper.getCurrentRowCount();
                data = movieMapper.getCurrentMovieAndPosterList(startIdx,moviePerPage);
            }

            totalPageNum = (int) Math.ceil((double) total/moviePerPage);
        } catch(IllegalArgumentException e){
            throw new IllegalArgumentException(INVALID_STATUS.getMessage());
        } catch (DataAccessException e){
            throw  new IllegalArgumentException(FAILED_BRING_MOVIES.getMessage());
        }

        PagingVO result = PagingVO.builder()
                .total(total)
                .totalPageNum(totalPageNum)
                .currentPage(currentPageNum)
                .data(data)
                .build();

        return result;
    }

```

nahokyun committed:
위 변경에 맞게 javadoc 수정

sunwootest commented on src/main/java/study/movieservice/service/MovieService.java:
```diff
@@ -38,4 +61,47 @@ public void addPoster(MultipartFile file, Long movieId) {
+     */
+    public PagingVO getMovieAndPosterList(Integer currentPageNum,String status){
+
+        int total=0;
+        int totalPageNum;
+        int startIdx = (currentPageNum - 1) * moviePerPage;
+        List<MovieAndPosterVO> data=null;
+
+        try{
+            MovieListType listStatus=MovieListType.valueOf(status);
+
+            if(listStatus==ALL) {
+                total = movieMapper.getTotalRowCount();
+                data = movieMapper.getAllMovieAndPosterList(startIdx,moviePerPage);
+            }
+            else if(listStatus==CURRENT){
+                total = movieMapper.getCurrentRowCount();
+                data = movieMapper.getCurrentMovieAndPosterList(startIdx,moviePerPage);
+            }
+
+            totalPageNum = (int) Math.ceil((double) total/moviePerPage);
```
삼항연산자를 사용하면 확실한 정보로 초기화할 수 있을 것 같네요 :)

nahokyun committed:
전체 영화목록, 상영중인 영화목록 불러오기#26

삼항연산자를 사용하여 total과 data값을 할당하는 부분을 간단히 표현하였습니다.

nahokyun commented:
> ### 변경된점
- 삼항연산자를 사용하여 total과 data값을 간단히 초기화 하도록 표현하였습니다.
## MovieService.java
### 변경 전 
```java
            if(listStatus==ALL) {
                total = movieMapper.getTotalRowCount();
                data = movieMapper.getAllMovieAndPosterList(startIdx,moviePerPage);
            }
            else if(listStatus==CURRENT){
                total = movieMapper.getCurrentRowCount();
                data = movieMapper.getCurrentMovieAndPosterList(startIdx,moviePerPage);
            }

```
### 변경 후 
```java
           total=(listStatus==ALL)?movieMapper.getTotalRowCount():movieMapper.getCurrentRowCount();
            data=(listStatus==ALL)?movieMapper.getAllMovieAndPosterList(startIdx,moviePerPage):movieMapper.getCurrentMovieAndPosterList(startIdx,moviePerPage);

```

sunwootest commented on src/main/java/study/movieservice/service/MovieService.java:
```diff
@@ -38,4 +61,41 @@ public void addPoster(MultipartFile file, Long movieId) {
+        try{
+            MovieListType listStatus=MovieListType.valueOf(status);
+
+            total=(listStatus==ALL)?movieMapper.getTotalRowCount():movieMapper.getCurrentRowCount();
+            data=(listStatus==ALL)?movieMapper.getAllMovieAndPosterList(startIdx,moviePerPage):movieMapper.getCurrentMovieAndPosterList(startIdx,moviePerPage);
```
코드가 너무 길어지면 개행을 넣어도 괜찮습니다.

sunwootest replied:
삼항연산자를 넣었으니 변수 초기화시에 자세한 정보로 초기화할 수 있겠네요.

nahokyun replied:
반영해 보도록 하겠습니다!

sunwootest commented on src/main/java/study/movieservice/service/MovieService.java:
```diff
@@ -38,4 +61,41 @@ public void addPoster(MultipartFile file, Long movieId) {
+            data=(listStatus==ALL)?movieMapper.getAllMovieAndPosterList(startIdx,moviePerPage):movieMapper.getCurrentMovieAndPosterList(startIdx,moviePerPage);
+
+            totalPageNum = (int) Math.ceil((double) total/moviePerPage);
+        } catch(IllegalArgumentException e){
+            throw new IllegalArgumentException(INVALID_STATUS.getMessage());
+        } catch (DataAccessException e){
+            throw  new IllegalArgumentException(FAILED_BRING_MOVIES.getMessage());
+        }
```
이 예외처리가 꼭 필요할까요?

nahokyun replied:
GeneralExceptionHandler에서 DataAccessException을 처리해줬기에 필요하지 않을것같고 enum값에 이상한 입력값이 왔을때 알려주는 IllegalArgumentException은 필요할것 같습니다

nahokyun committed:
전체 영화목록, 상영중인 영화목록 불러오기#26

불필요하게 긴 초기화부분을 삼항연산자를 사용함으로 인해 간소화 하고 이전에 처리한 예외처리를 사용하여 마찬가지로 불필요한 부분을 제거하였습니다.

nahokyun commented:
> ### 변경된점
- 이미 처리한 예외인 DataAccessException처리관련 내용을 삭제하였습니다.
- 삼항연산자를 사용함으로서 초기화를 한번에 진행하게하여 try-catch문을 간소화하였습니다.
## MovieService.java
### 변경 전 
```java
    public PagingVO getMovieAndPosterList(Integer currentPageNum,String status){

        int total=0;
        int totalPageNum;
        int startIdx = (currentPageNum - 1) * moviePerPage;
        List<MovieAndPosterVO> data=null;

        try{
            MovieListType listStatus=MovieListType.valueOf(status);

            total=(listStatus==ALL)?movieMapper.getTotalRowCount():movieMapper.getCurrentRowCount();
            data=(listStatus==ALL)?movieMapper.getAllMovieAndPosterList(startIdx,moviePerPage):movieMapper.getCurrentMovieAndPosterList(startIdx,moviePerPage);

            totalPageNum = (int) Math.ceil((double) total/moviePerPage);
        } catch(IllegalArgumentException e){
            throw new IllegalArgumentException(INVALID_STATUS.getMessage());
        } catch (DataAccessException e){
            throw  new IllegalArgumentException(FAILED_BRING_MOVIES.getMessage());
        }

        PagingVO result = PagingVO.builder()
                .total(total)
                .totalPageNum(totalPageNum)
                .currentPage(currentPageNum)
                .data(data)
                .build();

        return result;
    }

```
### 변경 후 
```java
    public PagingVO getMovieAndPosterList(Integer currentPageNum, String status) {

        int totalPageNum;
        int startIdx = (currentPageNum - 1) * moviePerPage;
        MovieListType listStatus;

        try {
            listStatus = MovieListType.valueOf(status);
        } catch (IllegalArgumentException e) {
            throw new IllegalArgumentException(INVALID_STATUS.getMessage());
        }

        int total = (listStatus == ALL) ? movieMapper.getTotalRowCount()
                : movieMapper.getCurrentRowCount();
        List<MovieAndPosterVO> data = (listStatus == ALL) ? movieMapper.getAllMovieAndPosterList(startIdx, moviePerPage)
                : movieMapper.getCurrentMovieAndPosterList(startIdx, moviePerPage);

        totalPageNum = (int) Math.ceil((double) total / moviePerPage);

        PagingVO result = PagingVO.builder()
                .total(total)
                .totalPageNum(totalPageNum)
                .currentPage(currentPageNum)
                .data(data)
                .build();

        return result;
    }

```

nahokyun committed:
Merge branch 'develop' into Movie-List#26
